File: backend/schedule-service/src/main/java/com/transport/urbain/service/BusService.java
package com.transport.urbain.service;

import com.transport.urbain.dto.request.CreateBusRequest;
import com.transport.urbain.dto.response.BusResponse;
import com.transport.urbain.model.BusStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * Service interface for bus management operations.
 * <p>
 * Provides methods for CRUD operations on buses, including creation, retrieval,
 * updating, deletion, and status management. Supports searching, filtering by status,
 * and accessing available buses for schedule assignment.
 */
public interface BusService {

    /**
     * Creates a new bus in the system.
     *
     * @param request the bus creation request containing bus details
     * @return the created bus response
     */
    BusResponse createBus(CreateBusRequest request);

    /**
     * Retrieves a bus by its unique identifier.
     *
     * @param id the bus ID
     * @return the bus response
     */
    BusResponse getBusById(Long id);

    /**
     * Retrieves a bus by its unique bus number.
     *
     * @param busNumber the bus number
     * @return the bus response
     */
    BusResponse getBusByNumber(String busNumber);

    /**
     * Retrieves all buses with pagination support.
     *
     * @param pageable pagination and sorting parameters
     * @return a page of bus responses
     */
    Page<BusResponse> getAllBuses(Pageable pageable);

    /**
     * Retrieves all available buses (active and in service).
     *
     * @return list of available bus responses
     */
    List<BusResponse> getAvailableBuses();

    /**
     * Retrieves all active buses with pagination support.
     *
     * @param pageable pagination and sorting parameters
     * @return a page of active bus responses
     */
    Page<BusResponse> getActiveBuses(Pageable pageable);

    /**
     * Retrieves buses filtered by status with pagination support.
     *
     * @param status the bus status to filter by
     * @param pageable pagination and sorting parameters
     * @return a page of buses with the specified status
     */
    Page<BusResponse> getBusesByStatus(BusStatus status, Pageable pageable);

    /**
     * Searches buses by keyword (bus number, license plate, or model).
     *
     * @param keyword the search keyword
     * @param pageable pagination and sorting parameters
     * @return a page of matching bus responses
     */
    Page<BusResponse> searchBuses(String keyword, Pageable pageable);

    /**
     * Updates an existing bus's information.
     *
     * @param id the bus ID to update
     * @param request the update request containing modified bus details
     * @return the updated bus response
     */
    BusResponse updateBus(Long id, CreateBusRequest request);

    /**
     * Deletes a bus from the system.
     *
     * @param id the bus ID to delete
     */
    void deleteBus(Long id);

    /**
     * Updates the status of a specific bus.
     *
     * @param id the bus ID
     * @param status the new status to apply
     */
    void updateBusStatus(Long id, BusStatus status);
}

File: frontend/src/services/busService.ts
import api from '@/lib/api'
import type { BusStatus } from '@/types/busStatus'

export interface BusResponse {
  id: number
  busNumber: string
  licensePlate: string
  model: string
  manufacturer: string
  year: number
  capacity: number
  seatingCapacity?: number
  standingCapacity?: number
  status: BusStatus
  hasWifi?: boolean
  hasAirConditioning?: boolean
  isAccessible: boolean
  hasGPS?: boolean
  lastMaintenanceDate?: string
  nextMaintenanceDate?: string
  notes?: string
  createdAt?: string
  updatedAt?: string
}

export interface CreateBusRequest {
  busNumber: string
  licensePlate: string
  model: string
  manufacturer: string
  year: number
  capacity: number
  seatingCapacity?: number
  standingCapacity?: number
  status: BusStatus
  hasWifi?: boolean
  hasAirConditioning?: boolean
  isAccessible?: boolean
  hasGPS?: boolean
  notes?: string
}

export interface PageResponse<T> {
  content: T[]
  totalElements: number
  totalPages: number
  size: number
  number: number
}

export const busService = {
  // Get all buses
  getAllBuses: async (page: number = 0, size: number = 10): Promise<PageResponse<BusResponse>> => {
    const response = await api.get<PageResponse<BusResponse>>('/api/buses', {
      params: { page, size },
    })
    return response.data
  },

  // Get bus by ID
  getBusById: async (id: number): Promise<BusResponse> => {
    const response = await api.get<BusResponse>(`/api/buses/${id}`)
    return response.data
  },

  // Create bus
  createBus: async (data: CreateBusRequest): Promise<BusResponse> => {
    const response = await api.post<BusResponse>('/api/buses', data)
    return response.data
  },

  // Update bus
  updateBus: async (id: number, data: Partial<CreateBusRequest>): Promise<BusResponse> => {
    const response = await api.put<BusResponse>(`/api/buses/${id}`, data)
    return response.data
  },

  // Delete bus
  deleteBus: async (id: number): Promise<void> => {
    await api.delete(`/api/buses/${id}`)
  },

  // Search buses (backend uses Pageable, not individual page/size params in search)
  searchBuses: async (keyword: string, page: number = 0, size: number = 10, sort: string = 'busNumber,asc'): Promise<PageResponse<BusResponse>> => {
    const response = await api.get<PageResponse<BusResponse>>('/api/buses/search', {
      params: { 
        keyword, 
        page, 
        size,
        sort
      },
    })
    return response.data
  },

  // Get buses by status
  getBusesByStatus: async (status: string, page: number = 0, size: number = 10): Promise<PageResponse<BusResponse>> => {
    const response = await api.get<PageResponse<BusResponse>>(`/api/buses/status/${status}`, {
      params: { page, size },
    })
    return response.data
  },

  // Get available buses (non-paginated list)
  getAvailableBuses: async (): Promise<BusResponse[]> => {
    const response = await api.get<BusResponse[]>('/api/buses/available')
    return response.data
  },

  // Update bus status
  updateBusStatus: async (id: number, status: string): Promise<void> => {
    await api.patch(`/api/buses/${id}/status`, null, {
      params: { status },
    })
  },

  // Get bus by bus number
  getBusByNumber: async (busNumber: string): Promise<BusResponse> => {
    const response = await api.get<BusResponse>(`/api/buses/number/${busNumber}`)
    return response.data
  },
}

